### DISCUSSION TITLE
How does the backend report errors?

### DISCUSSION THREAD

**Discussion by joulev** [joulev/ezkomment#101] [Questions] [answered]
Say, if there's something wrong with my request to the backend, and some errors happen (e.g. missing `Authorization`, missing `idToken`, body not formatted correctly etc.

How does the backend report that error back to me?

I suggest setting up a system of error codes in lieu of error messages as codes are much more concise.

**Comment by VietAnh1010** [comment, marked as answer]
We shall follow the normal [HTTP reponse status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) conventions.

When some resource is created / uploaded, 201 will be sent back. For other suceeded requests, 200 will be sent back.
For endpoints which is protected by JWT authentication, if the JWT token is invalid (no Bearer, missing charaters, etc.) then 403 will be sent back.
If a resource cannot be found, then 404 will be sent back.
If a request contains invalid data (for example: empty display name), then 400 will be returned.
If some unexcepted errors happened, they will be logged (only in production) and 500 will be sent back.
In development, all errors will also be logged to stderr using console.error (I am considering using console.dir with { depth: null } to log all properties of the errors).

**Reply by VietAnh1010** [reply]
I have written a custom error class `CustomApiError` to wrap all known errors.

```ts
export default class CustomApiError extends Error {
    public code: number;

    constructor(errOrMsg: string | Error, code: number = 400) {
        super(errOrMsg instanceof Error ? errOrMsg.message : errOrMsg);
        this.code = code;
    }
}
```

All errors will be handled by the `onError` property of `next-connect` (so I don't have to write multiple try-catch statements, just let the error propargates until it is handled by `next-connect`). If an error is an instance of `CustomApiError`, its error code and message will be extracted an sent back to the client.

Unknown error will be logged.

**Comment by VietAnh1010** [comment]


**Reply by joulev** [reply]
Hi, this is a Q/A discussion, so please edit your answer instead (or post it as a comment to your answer). I will delete this "answer" shortly.
